bind + argc specialization = 20x perf boost

xeno-by opened the pull request:
Default logic of mirror construction, which gets triggered via
reflectField/reflectMethod/reflectConstructor, validates a lot of facts
about its arguments.

This takes quite a bit of time, which significantly degrades performance
of reflection-heavy applications. Proposed two changes provide an order
of magnitude performance boost to a simple app, which repeatedly invokes
the same method for different receiver instances.

xeno-by committed:
bind + argc specialization = 20x perf boost

Default logic of mirror construction, which gets triggered via
reflectField/reflectMethod/reflectConstructor, validates a lot of facts
about its arguments.

This takes quite a bit of time, which significantly degrades performance
of reflection-heavy applications. Proposed two changes provide an order
of magnitude performance boost to a simple app, which repeatedly invokes
the same method for different receiver instances.

xeno-by commented:
review @retronym @gkossakowski

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1969/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1969/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1251/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1251/

gkossakowski commented:
@xeno-by: I can see how avoiding sanity checks and initialization can help with performance. I don't see how mirrors specialized on argument arity help. Can you comment on that?

xeno-by commented:
Apparently, `Seq.toArray` which is called by `_*` is slower than `new Array[Object]{args(0), ...}`. If my memory serves me right (it was a few months ago and, unfortunately, I don't have those notes anymore), the difference in performance starts from 4x with no arguments and becomes non-existent or even negative when argc is greater than 4.

gkossakowski commented:
The `Seq.toArray` requires `ClassManifest` for Array creation. I just checked that `Manifest[Any]` does not override `newArray` method, see:

https://github.com/scala/scala/blob/master/src/library/scala/reflect/Manifest.scala#L164

Maybe just overriding `newArray` there would help? Also, you can fix this issue locally in similar way as in fb31577. Try this and check if array allocation is the only culprit.

xeno-by commented:
```
import scala.reflect.runtime.universe._

class C {
  def foo = 2
}

object Test extends App {
  val cm = scala.reflect.runtime.currentMirror
  val mm = cm.reflect(new C).reflectMethod(typeOf[C].member(newTermName("foo")).asMethod)
  var i = 0
  while (i < 1000000000) {
    i += 1
    mm()
  }
}
```

With the proposed patch:

```
22:30 ~/Projects/Kepler_6638/sandbox (ticket/6638)$ time scala Test

real 0m7.813s
user 0m7.812s
sys  0m0.329s
```

Without argc specialization:

```
22:31 ~/Projects/Kepler_6638/sandbox (ticket/6638)$ time scala Test

real 0m30.185s
user 0m29.867s
sys  0m0.880s
```

Without argc specialization, but with overridden `newArray` in `Manifest.AnyRef`:

```
22:32 ~/Projects/Kepler_6638/sandbox (ticket/6638)$ time scala Test

real 0m17.519s
user 0m17.360s
sys  0m0.720s
```

paulp commented:
You reminded me I never got around to checking this in: [specialized-small-seqs](https://github.com/paulp/scala/tree/topic/specialized-small-seqs).

``` scala
class A {
  def f(xs: Int*) = xs.sum
  def g = f(101, 202, 303)
}
```

``` diff
diff --git i/A.class w/A.class
index d46753d3f3..a2d31ef551 100644
--- i/A.class
+++ w/A.class
@@ -19,21 +19,12 @@ public class A {
   public int g();
     Signature: ()I
     Code:
-        : getstatic     #                  // Field scala/Predef$.MODULE$:Lscala/Predef$;
-        : iconst_3
-        : newarray       int
+        : new           #                  // class scala/collection/immutable/SpecSeq Int
         : dup
         : bipush
-       : iastore
-       : dup
-       : iconst_1
+        : sipush
        : sipush
-       : iastore
-       : dup
-       : iconst_2
-       : sipush
-       : iastore
-       : invokevirtual #                  // Method scala/LowPriorityImplicits.wrapIntArray:([I)Lscala/collection/mutable/WrappedArray;
+       : invokespecial #                  // Method scala/collection/immutable/SpecSeq Int."<init>":(III)V
        : invokevirtual #                  // Method f:(Lscala/collection/Seq;)I
        : ireturn
```

gkossakowski commented:
Am I right that argc specialization combines two things:
1. Bypassing sanity checks
2. Specialization for arity

Right? If so, my suspicion was that specialization for arity is not needed because all it achieves is more efficient array allocation which can be fixed elsewhere without introducing 4 new classes. However, I think bypassing sanity checks is still a good idea (and probably we need to spend additional class on that, right).

Am I correct with all of that?

xeno-by committed:
a few performance improvements for toArray

First of all the typo I have made when migrating from manifests to tags.
`repr.getClass` in `WrappedArray` should read `array.getClass`.

Secondly manifests for Any, Object/AnyRef, AnyVal, Null and Nothing
now have their `newArray` methods overridden to avoid reflective overhead
of array instantiation.

xeno-by commented:
First of all, while looking into whether or not we can remove argc specialization (by which I meant specialization for arity) I found a bug which I introduced when migrating to tags. That's committed above. Hence PLS REBUILD ALL.

As to where we can avoid specialization, I'm not sure. The code that gets called resides in [1]. At first I thought that reflection performed to get the component type of an array might be at fault. But then I commented out everything in the body of toArray except for `array.asInstanceOf[Array[U]]` and it still was at about 17 seconds. Therefore it looks like `args.toArray(ClassTag.AnyRef)` is noticeably slower than `new Object[0]` even if it just calls `args.array`. How could we deal with that.

[1] https://github.com/scalamacros/kepler/blob/4ead04f46b524ac1e62a38dd6076265ff7daeee0/src/library/scala/collection/mutable/WrappedArray.scala#L67

paulp commented:
As a matter of general practice, when you're sweating the performance of a method, don't use a context bound and "implicitly", just pass the implicit parameter. It will never hurt and can easily help.

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1971/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1971/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1253/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1253/

paulp commented on src/reflect/scala/reflect/api/Mirrors.scala:
```diff
@@ -352,6 +352,11 @@ trait Mirrors { self: Universe =>
+    /** Creates a new mirror which uses the same symbol, but is bound to a different receiver.
+     *  This is significantly faster than recreating the mirror from scratch.
+     */
+    def bind(newReceiver: Any): FieldMirror
```
What do you think about "rebind" instead of "bind".

xeno-by replied:
To tell the truth, the original signature was `rebind :: Any => Unit`, but when I changed it to be non-mutating I thought that `bind` would fit better.

paulp commented on src/reflect/scala/reflect/runtime/JavaMirrors.scala:
```diff
@@ -351,11 +363,43 @@ private[reflect] trait JavaMirrors extends internal.SymbolTable with api.JavaUni
+      override def bind(newReceiver: Any) = new JavaVanillaMethodMirror4(newReceiver, symbol)
+      override def jinvokeraw(jmeth: jMethod, receiver: Any, args: Seq[Any]) = jmeth.invoke(receiver, args(0).asInstanceOf[AnyRef], args(1).asInstanceOf[AnyRef], args(2).asInstanceOf[AnyRef], args(3).asInstanceOf[AnyRef])
+    }
+
```
It would be nice to hit higher generality but as long as these are private classes I don't think we should worry about it overmuch.

paulp commented:
This looks good to me modulo my minor comments.
